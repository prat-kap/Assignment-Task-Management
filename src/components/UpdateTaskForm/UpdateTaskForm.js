import React, { useState } from "react"
import { useNavigate, useLocation } from "react-router-dom"
import { Form, Button, Container, Card, Alert } from "react-bootstrap"

import Header from "../Header/Header"
import Loading from "../Loader/Loading"

import { tasksData } from "../../services/tasks.services"

import { Formik } from "formik"
import {
  LABELS,
  MESSAGES,
  statuses,
  priorities
} from "../../constants/CommonConsts"

import "./UpdateTaskForm.css"

const UpdateTaskForm = () => {
  const {
    TITLE,
    DESCRIPTION,
    STATUS,
    SET_PRIORITY,
    EDIT_TASK,
    ADD_TASK,
    DUE_DATE,
    CREATE_UPDATE_TASK,
    PENDING
  } = LABELS
  const { NEW_TASK_ADDED, TASK_UPDATED } = MESSAGES

  const [message, setMessage] = useState({ error: false, msg: "" })
  const [loading, setLoading] = useState(false)

  const navigate = useNavigate()
  const location = useLocation()

  const isEdit = location.state?.id
  const taskDetails = location.state?.taskDetails

  const handleSubmit = async values => {
    try {
      if (!isEdit) {
        setMessage({
          error: false,
          msg: NEW_TASK_ADDED
        })
        setLoading(true)
        await tasksData.addTask(values)
      } else {
        setMessage({
          error: false,
          msg: TASK_UPDATED
        })
        setLoading(true)
        await tasksData.editTask(location.state?.id, values)
      }
      navigate("/")
    } catch (err) {
      setMessage({ error: true, msg: err.message })
    }
    setLoading(false)
  }
  return (
    <>
      <Header />
      {loading && <Loading />}
      <Container
        className="d-flex align-items-center justify-content-center"
        style={{ minHeight: "85vh" }}
      >
        <div className="w-100" style={{ maxWidth: "400px" }}>
          <Formik
            initialValues={{
              title: isEdit ? taskDetails.title : "",
              description: isEdit ? taskDetails.description : "",
              status: isEdit ? taskDetails.status : PENDING,
              dueDate: isEdit ? taskDetails.dueDate : new Date(),
              priority: isEdit ? taskDetails.priority : ""
            }}
            onSubmit={handleSubmit}
          >
            {({ handleSubmit, handleChange, values }) => (
              <>
                <Card>
                  <Card.Body>
                    <h2 className="text-center mb-4">{CREATE_UPDATE_TASK}</h2>
                    {message?.msg && (
                      <Alert
                        variant={message?.error ? "danger" : "success"}
                        dismissible
                        onClose={() => setMessage({ error: false, msg: "" })}
                      >
                        {message?.msg}
                      </Alert>
                    )}
                    <Form onSubmit={handleSubmit}>
                      <Form.Group controlId="title">
                        <Form.Label>{TITLE}</Form.Label>
                        <Form.Control
                          type="text"
                          name="title"
                          data-testid="title"
                          value={values.title}
                          onChange={handleChange}
                          required
                        />
                      </Form.Group>
                      <Form.Group controlId="description">
                        <Form.Label>{DESCRIPTION}</Form.Label>
                        <Form.Control
                          type="text"
                          name="description"
                          data-testid="description"
                          value={values.description}
                          onChange={handleChange}
                          required
                        />
                      </Form.Group>
                      <Form.Group controlId="dueDate">
                        <Form.Label>{DUE_DATE}</Form.Label>
                        <Form.Control
                          type="date"
                          name="dueDate"
                          data-testid="due-date"
                          min={new Date().toISOString().split("T")[0]}
                          value={values.dueDate}
                          onChange={handleChange}
                          required
                        />
                      </Form.Group>
                      <Form.Group controlId="status">
                        <Form.Label>{STATUS}</Form.Label>
                        <Form.Control
                          type="text"
                          name="status"
                          as="select"
                          data-testid="status"
                          value={values.status}
                          onChange={handleChange}
                          className="custom_select"
                          required
                        >
                          {statuses.map(item => (
                            <option key={item.key} value={item.value}>
                              {item.value}
                            </option>
                          ))}
                        </Form.Control>
                      </Form.Group>
                      <Form.Group controlId="priority">
                        <Form.Label>{SET_PRIORITY}</Form.Label>
                        <Form.Control
                          type="text"
                          name="priority"
                          as="select"
                          data-testid="priority"
                          value={values.priority}
                          onChange={handleChange}
                          className="custom_select"
                          required
                        >
                          {priorities.map(item => (
                            <option key={item.key} value={item.value}>
                              {item.value}
                            </option>
                          ))}
                        </Form.Control>
                      </Form.Group>
                      <Button
                        disabled={loading}
                        className="w-100 mt-3"
                        type="submit"
                        name="submit"
                        data-testid="submit-btn"
                      >
                        {isEdit ? EDIT_TASK : ADD_TASK}
                      </Button>
                    </Form>
                  </Card.Body>
                </Card>
              </>
            )}
          </Formik>
        </div>
      </Container>
    </>
  )
}

export default UpdateTaskForm
