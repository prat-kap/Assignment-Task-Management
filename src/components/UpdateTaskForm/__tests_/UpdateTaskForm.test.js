import { BrowserRouter as Router } from "react-router-dom"
import { render, screen } from "@testing-library/react"
import userEvent from "@testing-library/user-event"

import UpdateTaskForm from "../UpdateTaskForm"
import Loading from "../../Loader/Loading"
import "../../../i18n"

test("Rendering and submitting add and update task form", async () => {
  const handleSubmit = jest.fn()
  render(
    <Router>
      <UpdateTaskForm onSubmit={handleSubmit} />
    </Router>
  )
  const user = userEvent.setup()

  user.type(screen.getByTestId("title", { name: "title" }), "Read a book")
  user.type(screen.getByTestId("description", { name: "title" }), "Self time")
  user.type(screen.getByTestId("due-date", { name: "dueDate" }), "22-09-2023")
  user.type(screen.getByTestId("status", { name: "status" }), "pending")
  user.type(screen.getByTestId("priority", { name: "priority" }), "1")

  user.click(screen.getByTestId("submit-btn", { name: "submit" }))
  const { container } = render(<Loading />)
  expect(container).toMatchSnapshot()
})
