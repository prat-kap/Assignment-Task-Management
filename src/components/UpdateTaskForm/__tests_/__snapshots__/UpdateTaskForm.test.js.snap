// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rendering and submitting add and update task form 1`] = `
<div>
  <div
    style="display: flex; justify-content: center; align-items: center; width: 100%; height: 100%;"
  >
    <div
      class="spinner-border"
      style="width: 80px; height: 80px;"
    />
  </div>
</div>
`;

exports[`rendering and submitting add and update task form 1`] = `
<div>
  <div
    style="display: flex; justify-content: center; align-items: center; width: 100%; height: 100%;"
  >
    <div
      class="spinner-border"
      style="width: 80px; height: 80px;"
    />
  </div>
</div>
`;
